src/Page.php

```diff
@@ -0,0 +1,169 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\GAX;
+
+use IteratorAggregate;
+
+/**
+ * A Page object wraps an API list method response and provides methods
+ * to retrieve additional pages using the page token.
+ */
+class Page implements IteratorAggregate
+{
+    const FINAL_PAGE_TOKEN = "";
+
+    private $parameters;
+    private $callable;
+    private $pageStreamingDescriptor;
+
+    private $pageToken;
+
+    private $response;
+
+    public function __construct($params, $callable, $pageStreamingDescriptor) {
+        if (empty($params) || !is_object($params[0])) {
+            throw new InvalidArgumentException('First argument must be a request object.');
+        }
+        $this->parameters = $params;
+        $this->callable = $callable;
+        $this->pageStreamingDescriptor = $pageStreamingDescriptor;
+
+        $requestPageTokenField = $this->pageStreamingDescriptor->getRequestPageTokenField();
+        $this->pageToken = $params[0]->$requestPageTokenField;
+
+        // Make gRPC call eagerly
+        $this->response = call_user_func_array($this->callable, $this->parameters);
+    }
+
+    /**
+     * Returns true if there are more pages that can be retrieved from the
+     * API.
+     */
+    public function hasNextPage() {
+        return strcmp($this->getNextPageToken(), Page::FINAL_PAGE_TOKEN) != 0;
+    }
+
+    /**
+     * Returns the next page token from the response.
+     */
+    public function getNextPageToken() {
+        $responsePageTokenField = $this->pageStreamingDescriptor->getResponsePageTokenField();
+        return $this->getResponseObject()->$responsePageTokenField;
+    }
+
+    /**
+     * Retrieves the next Page object using the next page token.
+     */
+    public function getNextPage($pageSize = null) {
```

Comment:
I am not opposed to this idea. The page_size option is available in the original API call as part of the optional parameters, so removing it from here won't lose much. And if the user really needs to retrieve a sequence of pages using different page_size options, they can manually manipulate page tokens and make API calls